### ISSUE TITLE
Improve mumble accessibility: Upgrade qt 5.12 LTS

### ISSUE THREAD

**Issue by dingpengyu** [mumble-voip/mumble#3808]
hi davidebeatrici
I checked mumble's accessibility carefully and found that mumble is not friendly to screen readers. Most of mumble's edit boxes cannot interact with the contents of the edit box using screen readers.
Based on the above question, can the official upgrade the mumble version of qt to 5.12 LTS?
Written in the official qt log, 5.12 LTS rewrites the accessibility module
The following is the update log page for the official qt 5.12 LTS:
[New Features in Qt 5.12](https://wiki.qt.io/New_Features_in_Qt_5.12)
Qt GUI
Added support to Windows UI Automation to the WinRT QPA, allowing Qt-based UWP applications to operate with accessibility and programmatic UI control tools.
thanks

**Comment by davidebeatrici** [comment]
Hi,

Mumble is a desktop application, not UWP.

Qt has a dedicated interface we can use, called `QAccessibleInterface`: https://doc.qt.io/qt-5/accessible-qwidget.html

> To provide accessibility support for a widget or other user interface element, you need to implement the [QAccessibleInterface](https://doc.qt.io/qt-5/qaccessibleinterface.html) and distribute it in a [QAccessiblePlugin](https://doc.qt.io/qt-5/qaccessibleplugin.html). It is also possible to compile the interface into the application and provide a [QAccessible::InterfaceFactory](https://doc.qt.io/qt-5/qaccessible.html#InterfaceFactory-typedef) for it. The factory can be used if you link statically or do not want the added complexity of plugins. This can be an advantage if you, for instance, are delivering a 3-rd party library.
>
> All widgets and other user interface elements should have interfaces and plugins. If you want your application to support accessibility, you will need to consider the following:
>
> - Qt already implements accessibility for its own widgets. We therefore recommend that you use Qt widgets where possible.
> - A [QAccessibleInterface](https://doc.qt.io/qt-5/qaccessibleinterface.html) needs to be implemented for each element that you want to make available to accessibility clients.
> - You need to send accessibility events from the custom user interface elements that you implement.

@Kissaki Any thoughts on it?

**Comment by dingpengyu** [comment]
> Hi,
> 
> Mumble is a desktop application, not UWP.
> 
> Qt has a dedicated interface we can use, called `QAccessibleInterface`: https://doc.qt.io/qt-5/accessible-qwidget.html
> 
> > To provide accessibility support for a widget or other user interface element, you need to implement the [QAccessibleInterface](https://doc.qt.io/qt-5/qaccessibleinterface.html) and distribute it in a [QAccessiblePlugin](https://doc.qt.io/qt-5/qaccessibleplugin.html). It is also possible to compile the interface into the application and provide a [QAccessible::InterfaceFactory](https://doc.qt.io/qt-5/qaccessible.html#InterfaceFactory-typedef) for it. The factory can be used if you link statically or do not want the added complexity of plugins. This can be an advantage if you, for instance, are delivering a 3-rd party library.
> > All widgets and other user interface elements should have interfaces and plugins. If you want your application to support accessibility, you will need to consider the following:
> > 
> > * Qt already implements accessibility for its own widgets. We therefore recommend that you use Qt widgets where possible.
> > * A [QAccessibleInterface](https://doc.qt.io/qt-5/qaccessibleinterface.html) needs to be implemented for each element that you want to make available to accessibility clients.
> > * You need to send accessibility events from the custom user interface elements that you implement.
> 
> @Kissaki Any thoughts on it?
Thank you for your information
Is it possible to do mumble's accessibility optimization?
thanks

**Comment by davidebeatrici** [comment]
It's definitely possible.

**Comment by dingpengyu** [comment]
> It's definitely possible.

thank you for your reply

In addition, if you can, you can also: Expose accessibility strings in shortcuts tree
#2293

**Comment by davidebeatrici** [comment]
You're welcome, I will also take a look at #2293.

**Comment by ethindp** [comment]
Are there any updates to this? I'm irritated that this project is still using out-of-date, deprecated, and probably insecure software libraries. I would build a client using QT5 if possible but its not right now. :(

**Comment by ethindp** [comment]
Could one of you produce a dynamically linked Mumble binary package for me? I would do so, but I'm not wasting space to install VS 2017 just to use it -- I'll stick with VS 2019. Unfortunately, Mumble does not support VS 2019 (for some reason).

**Comment by davidebeatrici** [comment]
What kind of issues are you encountering with Visual Studio 2019?

**Comment by ethindp** [comment]
If I run setup-x64.cmd it builds the env, but when I run the shortcut it gives me a few errors about something not being found. (I attribute it to the prep script nto supporting VS 2019 at all.) When I run the BASH script, it bails out claiming that an executable or something couldn't be found almost immediately.

**Comment by davidebeatrici** [comment]
It's our build environment that is incompatible with Visual Studio 2019, not Mumble itself.

We are working on a CMake project, I will update my branch with the latest changes and let you know.

**Comment by ethindp** [comment]
Could you produce for mea dynamically linked binary? I'd like to see if NVDA issues persist in dynamically linked setups.

**Comment by davidebeatrici** [comment]
What are the issues you're encountering with NVDA?

**Comment by ethindp** [comment]
NVDA locks up and does not respond (i.e. stop talking) when any keys
are pressed, and I am unable to send keyboard commands to it until I
make the application lose focus. I am pretty sure that this is
something you guys are doing on your end since I am unable to
reproduce it in any other QT application.

On 11/13/19, Davide Beatrici <notifications@github.com> wrote:
> What are the issues you're encountering with NVDA?
>
> --
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/mumble-voip/mumble/issues/3808#issuecomment-553644645


-- 
Signed,
Ethin D. Probst

**Comment by davidebeatrici** [comment]
Could you test whether the problem still occurs after disabling the Windows hooks?
https://dl.mumble.info/reg/disable-winhooks.reg

To enable them again: https://dl.mumble.info/reg/enable-winhooks.reg

**Comment by ethindp** [comment]
Nope, it stopped. I think that this should be default behavior if a
screen reader is detected -- disable all windows hooks. (What are the
hooks even used for?)

On 11/13/19, Davide Beatrici <notifications@github.com> wrote:
> Could you test whether the problem still occurs after disabling the Windows
> hooks?
> https://dl.mumble.info/reg/disable-winhooks.reg
>
> To enable them again: https://dl.mumble.info/reg/enable-winhooks.reg
>
> --
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/mumble-voip/mumble/issues/3808#issuecomment-553647818


-- 
Signed,
Ethin D. Probst

**Comment by davidebeatrici** [comment]
The hooks are used for the global shortcuts:
https://github.com/mumble-voip/mumble/blob/2004de970cb66e511d7f39f6752c386e5bd86f61/src/mumble/GlobalShortcut_win.cpp#L764-L796
